torch/csrc/CudaIPCTypes.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+#ifdef USE_CUDA
+#include <c10/core/Allocator.h>
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <c10/cuda/CUDAException.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/cuda/CUDAStream.h>
+#include <c10/util/Logging.h>
+#include <cuda_runtime_api.h>
+#include <iostream>
+
+namespace torch {
+
+constexpr uint CUDA_IPC_REF_COUNTER_FILE_SIZE = 10000;
+constexpr uint CUDA_IPC_WARN_AFTER_X_BLOCKS_IN_LIMBO = 1000;
+
+void CudaIPCCollect();
+
+at::DataPtr GetNewRefCountedSentData(void* data, at::Device device);
+
+bool CudaIPCHaveRefCounter();
+void CudaIPCCreateRefCounter(
+    std::string handle,
+    uint64_t size,
+    at::DataPtr data_ptr);
+void ReturnRefCounter(std::string handle, uint64_t offset);
+
+struct CudaIPCSentData final {
+  std::string handle_;
+  int64_t offset_;
+  int64_t* counter_ptr_; // Reference counter shared memory block
+  at::DataPtr original_ptr_; // Original mem allocation
+  cudaEvent_t event_; // Sync cuEventDestroy
+  at::Device device_;
+
+  CudaIPCSentData(
+      std::string handle,
+      int64_t offset,
+      int64_t* counter_ptr,
+      at::Device device)
+      : handle_(handle),
+        offset_(offset),
+        counter_ptr_(counter_ptr),
+        original_ptr_(),
+        device_(device) {
+#ifndef __HIP_PLATFORM_HCC__
+    // TODO: More efficient would be to create event inside of main thread (at
+    // the moment of the queue.put)
```

Comment:
"The reason this is more efficient is because the main thread may have queued extra work on the stream, which this event will consequently wait for (uselessly)."

Reply:
This TODO is sort of in the wrong place, because you couldn't fix this problem by editing the CudaIPCSentData constructor; you'd have to fix the call site.